### ISSUE TITLE
flow matching loss objective sign?

### ISSUE THREAD

**Issue by YeonjeeJung** [lucidrains/pi-zero-pytorch#14]
Hello, thank you very much for your wonderful code implementation.

I have one small question about the flow matching loss.
In this paper, the denoising vector field u is defined as epsilon-action_t.
But I found out that you define the flow by action - noise.
This is thought to be the concept of the opposite sign.

Can I get an explanation for this implementation?

**Comment by lucidrains** [comment]
@YeonjeeJung ah i see, so they do

i don't think the sign matters all that much to be honest. and as for using `action_t` instead of `action`, well, that's just a scale away that the network should learn quite easily given the time conditioning

**Comment by lucidrains** [comment]
afaik, the predicted flow should be `x1 - x0`, feel free to point me at some flow literature that states otherwise

**Comment by lucidrains** [comment]
@YeonjeeJung oh just took a second look and noticed that A superscript t is their action at time t (so confusing)

ok, so it is equivalent besides the sign, which is trivial!

**Comment by lucidrains** [comment]
@YeonjeeJung decided to just make it line up with the paper, even if it matters very little

**Comment by YeonjeeJung** [comment]
Oh thank you for correction!
But... I have same opinion with you that the objective should be x1-x0.

> afaik, the traditional rectified flow matching objective should be `x1 - x0`, feel free to point me at some flow literature that states otherwise, or perhaps have experiments showing one way is better than the other

I'll think about it more and share it with you.

And, I have another questions in the code.
It is thought to be use 'sample_actions' function in inference time.
In ODE_fn, inside the forward_with_reward_cfg, there are two forwards, with reward and without reward.
What is the purpose of the reward, and why was it used only for inference time?

**Comment by lucidrains** [comment]
@YeonjeeJung just ignore the reward tokens for now

that's just my ongoing attempt to extend this work beyond just behavioral cloning / imitation learning

**Comment by lucidrains** [comment]
> Oh thank you for correction! But... I have same opinion with you that the objective should be x1-x0.

oh, they did have it correct, just that their notation is weird. their action at time `t` has both sub and superscript `t`, so their equation comes out to `x0 - x1`

**Comment by Wonder1905** [comment]
> @YeonjeeJung just ignore the reward tokens for now
> 
> that's just my ongoing attempt to extend this work beyond just behavioral cloning / imitation learning

Isnt the call for the two forward be the other way around?
`        action_flow_with_reward, with_reward_cache_kv = self.forward(
            *args,
            reward_tokens = reward_tokens,
            cached_state_keys_values = with_reward_cache,
            **forward_kwargs,
            **kwargs
        )

        if not exists(reward_tokens) or cond_scale == 0.:
            return action_flow_with_reward, (with_reward_cache_kv, None)

        assert self.can_cfg, 'you need to train with reward token dropout'

        action_flow_without_reward, without_reward_cache_kv = self.forward(
            *args,
            cached_state_keys_values = without_reward_cache,
            **forward_kwargs,
            **kwargs
        )`
        
If reward_tokens are None, we should on do forward without rewards and return.
Otherwise run also with rewards.

Im just trying to understand how sample_actions would be kept legit without using reward_tokens.
But there is a chance I'm missing something.

**Comment by lucidrains** [comment]
@Wonder1905 if reward tokens is None, then the first call is without reward conditioning, which I then early return
